python/lsst/daf/ingest/ingestCatalogTask.py

```diff
@@ -0,0 +1,580 @@
+    if sz > 65535:
+        # If necessary, longer strings could be ingested as TEXT.
+        raise RuntimeError("String field is too large for ingestion")
+    elif sz == 0:
+        raise RuntimeError("String field has zero size")
+    # A string containing trailing spaces cannot round-trip to a CHAR
+    # column and back. Therefore, use VARCHAR. Also, assume strings are
+    # ASCII for now.
+    return "VARCHAR({}) CHARACTER SET ascii COLLATE ascii_bin NOT NULL".format(sz)
+
+def _sqlTypeForArray(formatChar, field):
+    """Helper function for computing the SQL column type of an array valued field.
+    See https://docs.python.org/2/library/struct.html#format-characters for possible
+    format characters. The formatChar passed in must correspond to the array element
+    type.
+    """
+    sz = field.getSize()
+    if sz == 0:
+        return "BLOB NOT NULL"
+    sz *= struct.calcsize("<" + formatChar)
+    if sz > 65535:
+        raise RuntimeError("Array field is too large for ingestion")
+    return "BINARY({}) NOT NULL".format(sz)
+
+
+"""A dictionary mapping afw field type strings to corresponding formatters.
+
+This dictionary is used by IngestCatalogTask to determine how to format each
+type of field in the input catalog.  If new field types are added to the afw
+table library, they should also be added here.
+"""
+fieldFormatters = dict(
+    U = FieldFormatter(lambda f: "SMALLINT UNSIGNED NOT NULL", lambda v: str(v)),
+    I = FieldFormatter(lambda f: "INT NOT NULL", lambda v: str(v)),
+    L = FieldFormatter(lambda f: "BIGINT NOT NULL", lambda v: str(v)),
+    F = FieldFormatter(lambda f: "FLOAT", lambda v: _formatNumber("{:.9g}", v)),
+    D = FieldFormatter(lambda f: "DOUBLE", lambda v: _formatNumber("{:.17g}", v)),
+    Flag = FieldFormatter(lambda f: "BIT NOT NULL", lambda v: "1" if v else "0"),
+    Angle = FieldFormatter(lambda f: "DOUBLE", lambda v: _formatNumber("{:.17g}", v.asDegrees())),
+    String = FieldFormatter(_sqlTypeForString, _formatString),
+    ArrayU = FieldFormatter(lambda f: _sqlTypeForArray("H", f), lambda v: _formatArray("H", v)),
+    ArrayI = FieldFormatter(lambda f: _sqlTypeForArray("i", f), lambda v: _formatArray("i", v)),
+    ArrayF = FieldFormatter(lambda f: _sqlTypeForArray("f", f), lambda v: _formatArray("f", v)),
+    ArrayD = FieldFormatter(lambda f: _sqlTypeForArray("d", f), lambda v: _formatArray("d", v)),
+)
+
+
+def canonicalizeFieldName(fieldName):
+    """Return a MySQL-compatible version of the schema field name, for now by
+    changing any non-word characters to underscores.
+    """
+    return re.sub(r"[^\w]", "_", fieldName)
+
+def quoteIdentifier(identifier):
+    """Return a MySQL compatible version of the given identifier by quoting
+    it with back-ticks (and escaping embedded back-ticks).
+    """
+    return "`" + identifier.replace("`", "``") + "`"
+
+def aliasesFor(name, mappings):
+    """Compute and return the set of possible aliases for the given field name.
```

Comment:
I'm conflicted about this.

First of all, I think the approach you've adopted here is basically correct given the way ``afw::table`` works. I think this is probably the best we're going to do for now, and I don't think we should delay merging this code and closing the ticket before moving on.

However, it seems like this is a problem with the implementation of aliases in ``afw``, and I wonder if, in the longer term, we ought to address that (or, at least, discuss the implications further with people like @TallJimbo). I'm pretty sure it's not a design goal that aliases like `base_CircularApertureFlux_flag_badCentroid_badInitialCentroid_badInitialCentroid_badInitialCentroid_badInitialCentroid_badInitialCentroid` should be supported, and, while they don't do any harm from an `afw` point of view, the complexity they're adding here seems like a bad thing.

Actually, I think there is a wrinkle with this from point of view of the pure-`afw` user who never interacts with aliases from a database point of view. Since the cycle detection scheme `afw` uses depends on the size of the alias map, whether or not a big alias like the one above ultimately resolves to anything useful can depend on how many otherwise completely unrelated aliases have been added to the map. Remove something ostensibly unrelated -- like, say, `slot_ModelFlux` -- and suddenly your `badInitialCentroid` alias starts throwing exceptions.

My suggestion would be to merge this as is (unless a change to the `afw` behaviour is imminent), but file a ticket against `afw` to discuss cleaning this up later.